File: python3/day_16_ Validate IP Address.py
class Solution:
    def validIPAddress(self, IP: str) -> str:
        def ip_v4(IP):
            dots = 0
            prev = '.'
            for i, curr in enumerate(IP):
                if prev == '.':
                    if curr == '0' and i+1 < len(IP) and IP[i+1] >= '0' and IP[i+1] <= '9':
                        return False
                    num = 0
                
                if curr == '.':
                    if i+1 < len(IP) and IP[i+1] == '.':
                        return False
                    dots += 1
                    
                    if num > 255 or num < 0:
                        return False
                else:
                    if curr >= '0' and curr <= '9':
                        num = num * 10 + int(curr)
                    else:
                        return False
                
                if num > 255 or num < 0:
                    return False
                prev = curr
                
            if num > 255 or num < 0:
                return False
            return True if dots == 3 and IP[-1] != '.' else False
        
        def ip_v6(IP):
            cols = 0
            prev = ':'
            for i, curr in enumerate(IP):
                if prev == ':':
                    num = 0
                
                if curr == ':':
                    if i+1 < len(IP) and IP[i+1] == ':':
                        return False
                    cols += 1
                    
                    if num > 4 or num < 0:
                        return False
                else:
                    if curr >= 'A' and curr <= 'F' or curr >= 'a' and curr <= 'f' or curr >= '0' and curr <= '9':
                        num += 1
                    else:
                        return False
                
                if num > 4 or num < 0:
                    return False
                prev = curr
                
            if num > 4 or num < 0:
                return False
            return True if cols == 7 and IP[-1] != ':' else False
            
        if '.' in IP and ':' in IP:
            return "Neither"
        elif '.' in IP and ip_v4(IP):
            return "IPv4"
        elif ':' in IP and ip_v6(IP):
            return "IPv6"
        return "Neither"

File: cpp/june/day_16_ Validate IP Address.cpp
class Solution {
public:
    bool ip_v4(string ip){
        int i, j, num, dots = 0;
        char prev, curr;
        // cout << ip << " :";
        prev = '.';
        for(i = 0; i < ip.size(); i++){
            curr = ip[i];
            if(prev == '.'){
                if(curr == '0' && i+1 < ip.size() && ip[i+1] >= '0' && ip[i+1] <= '9'){
                    return false;
                }
                num = 0;
            }

            if(curr == '.'){
                if(i+1 < ip.size() && ip[i+1] == '.'){
                    return false;
                }
                dots++;
                // cout << num << " ";
                if(num > 255 || num < 0){
                    return false;
                }
            }else{
                if(curr >= '0' && curr <= '9'){
                    num = num * 10 + curr - '0';
                }else{
                    return false;
                }
                
            }
            if(num > 255 || num < 0){
                return false;
            }
            
            prev = curr;
        }
        // cout <<endl;
        if(num > 255 || num < 0){
            return false;
        }
        return dots == 3 && ip[ip.size()-1] != '.' ? true : false;
    }
    
    bool ip_v6(string ip){
        int i, j, num, cols = 0;
        char prev, curr;
        // cout << ip << " :";
        prev = ':';
        for(i = 0; i < ip.size(); i++){
            curr = ip[i];
            if(prev == ':'){
                // if(curr == '0' && i < ip.size() && ip[i+1] >= '0' && ip[i+1] <= '9'){
                //     return false;
                // }
                num = 0;
            }

            if(curr == ':'){
                if(i+1 < ip.size() && ip[i+1] == ':'){
                    return false;
                }
                cols++;
                // cout << num << " ";
                if(num > 4 || num < 0){
                    return false;
                }
            }else{
                // num = num * 10 + curr - '0';
                if(curr >= 'A' && curr <= 'F' || curr >= 'a' && curr <= 'f' || curr >= '0' && curr <= '9'){
                    num++;
                }else{
                    return false;
                }
                
            }
            if(num > 4 || num < 0){
                return false;
            }
            prev = curr;
        }
        // cout <<endl;
        if(num > 4 || num < 0){
            return false;
        }
        return cols == 7 && ip[ip.size()-1] != ':'  ? true : false;
    }
    
    string validIPAddress(string IP) {
        // cout << endl;
        long long d = IP.find('.');
        long long c = IP.find(':');
        // cout << d << endl;
        if(d > -1 && c == -1 && ip_v4(IP)){
            return "IPv4";
        }else if(d == -1 && c > -1 && ip_v6(IP)){
            return "IPv6";
        }
        return "Neither";
    }
};
